src/napari/_vispy/canvas.py

```diff
@@ -875,41 +903,109 @@ def _update_layer_overlays(self, layer: Layer) -> None:
 
             self._add_layer_overlay(layer, overlay, parent)
 
-        self._update_overlay_canvas_positions()
+        self._defer_overlay_position_update()
 
     def _get_ordered_visible_canvas_overlays(self):
         # note that some canvas overlays do no use CanvasPosition, but are instead
         # free-floating (such as the cursor overlay), so those are skipped
 
-        # first viewer overlays
+        # first the base view (non-gridded viewer overlays)
         for overlay, vispy_overlays in self._overlay_to_visual.items():
             if (
                 overlay.visible
                 and isinstance(overlay, CanvasOverlay)
                 and overlay.position in list(CanvasPosition)
+                and not (overlay.gridded and self.viewer.grid.enabled)
             ):
-                yield overlay, vispy_overlays
+                yield overlay, vispy_overlays[0], None
 
-        # then layer overlays
-        for layer in self.viewer.layers:
-            for overlay, vispy_overlay in self._layer_overlay_to_visual.get(
-                layer, {}
-            ).items():
+        # then gridded viewer overlays and layer overlays, by viewbox, in order
+        for viewbox_idx, (_, layer_indices) in enumerate(
+            self.viewer.grid.iter_viewboxes(len(self.viewer.layers))
+        ):
+            if not layer_indices:
+                # last empty boxes of the grid
+                break
+            view = viewbox_idx if self.viewer.grid.enabled else None
```

Comment:
I see that here is an attempt to generalize code for grid and non-grid mode, but it makes it harder to follow for me. 

As I understand it, if the grid mode is disabled, the for loop should do only one iteration?

Reply:
Why? I think it makes sense that if you're itering the viewboxes, you get all of them and some might be empty, no? It's not `iter_layers`